rotkehlchen/chain/bitcoin/transactions.py

```diff
@@ -0,0 +1,226 @@
+        base_url=BLOCKSTREAM_BASE_URL,
+        url_suffix=url_suffix,
+    )
+
+
+def _query_mempool_space_txs(url_suffix: str):
+    return _query_blockstream_mempool_txs(
+        base_url=MEMPOOL_SPACE_BASE_URL,
+        url_suffix=url_suffix,
+    )
+
+
+def _query_tx_list_from_api(
+        account: BTCAddress,
+        from_timestamp: Timestamp,
+        to_timestamp: Timestamp,
+        existing_tx_id: str | None,
+) -> list[BitcoinTx]:
+    """Query raw bitcoin tx list from the block explorer apis and deserialize them.
+    The txs are ordered with newest first from the api, with results automatically paginated.
+    The next page is requested by specifying the last seen tx id from the previous result.
+
+    If `existing_tx_id` is specified, querying will stop when that id is encountered, only
+    returning new transactions since that id.
+
+    Transactions newer than `to_timestamp` will be skipped, but `from_timestamp` is currently
+    ignored - since the api returns newest to oldest, if we quit before querying to the end, the
+    next query will stop at `existing_tx_id` and the skipped older txs will never be queried.
+
+    Returns a list of deserialized BitcoinTxs.
+    """
+    tx_list, tx_id, to_timestamp_ms = [], None, ts_sec_to_ms(to_timestamp)
+    tx_endpoint = f'address/{account}/txs/chain'
+    while True:
+        for entry in (raw_tx_list := query_apis_via_callbacks(
+            api_callbacks={
+                'blockstream.info': _query_blockstream_txs,
+                'mempool.space': _query_mempool_space_txs,
+            },
+            url_suffix=f'{tx_endpoint}/{tx_id}' if tx_id is not None else tx_endpoint,  # Specify the last seen tx id to request the next page
+        )):
+            tx = deserialize_bitcoin_tx(entry)
+            if tx.timestamp > to_timestamp_ms:
+                continue  # Haven't reached the requested range yet. Skip tx.
+            if (tx_id := tx.tx_id) == existing_tx_id:
+                return tx_list  # All new txs have been queried. Skip tx and return.
+
+            tx_list.append(tx)
+
+        if len(raw_tx_list) < 25:
+            break
+
+    return tx_list
+
+
+def decode_history_events_for_bitcoin_tx(tx: BitcoinTx) -> list[HistoryEvent]:
+    """Decode a BitcoinTx into HistoryEvents
+
+    If there are multiple inputs/outputs for the same address, events will only be created
+    with the total in amount and total out amount for that address.
+
+    If there are inputs from multiple addresses a fee event will be created for each address
+    with the fee amounts proportional to the amounts spent by each address.
+
+    Returns a list of HistoryEvents
+    """
+    events, sequence_index = [], -1
+    for tx_in_out_list, event_type in (
+            (tx.inputs, HistoryEventType.SPEND),
+            (tx.outputs, HistoryEventType.RECEIVE),
+    ):
+        per_address_totals = defaultdict(lambda: ZERO)
+        for tx_in_out in tx_in_out_list:
```

Comment:
i think you need to account for self transfers i.e. an account is consolidating UXTOs. the easiest way might be to check the net amount (spend - receive). in such cases, it should probably be an informational event.